File: 037_ReorderList.cpp
// Topic: Reorder List
// Description: Rearrange nodes: first, last, second, second-last pattern.
// Approach: Find middle, reverse second half, and merge alternatingly.

#include <bits/stdc++.h>
using namespace std;

struct ListNode
{
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
};

void reorderList(ListNode* head)
{
    if (!head || !head->next)
        return;
    ListNode *slow = head, *fast = head;
    while (fast->next && fast->next->next)
    {
        slow = slow->next;
        fast = fast->next->next;
    }
    ListNode* second = slow->next;
    slow->next = nullptr;
    // reverse second
    ListNode* prev = nullptr;
    while (second)
    {
        ListNode* nxt = second->next;
        second->next = prev;
        prev = second;
        second = nxt;
    }
    // merge
    ListNode* p = head;
    ListNode* q = prev;
    while (q)
    {
        ListNode* pn = p->next;
        ListNode* qn = q->next;
        p->next = q;
        q->next = pn;
        p = pn;
        q = qn;
    }
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    // Test 1
    {
        ListNode* h = new ListNode(1);
        h->next = new ListNode(2);
        h->next->next = new ListNode(3);
        h->next->next->next = new ListNode(4);
        reorderList(h);
        for (ListNode* p = h; p; p = p->next)
        {
            cout << p->val;
            if (p->next)
                cout << " ";
        }
        cout << "\n";
    }
    // Test 2
    {
        ListNode* h = new ListNode(1);
        h->next = new ListNode(2);
        h->next->next = new ListNode(3);
        h->next->next->next = new ListNode(4);
        h->next->next->next->next = new ListNode(5);
        reorderList(h);
        for (ListNode* p = h; p; p = p->next)
        {
            cout << p->val;
            if (p->next)
                cout << " ";
        }
        cout << "\n";
    }
}

/*
Time Complexity: O(n)
Space Complexity: O(1)
*/

File: 037_ReorderList.py
# Topic: Reorder List
# Description: Rearrange nodes: first, last, second, second-last pattern.
# Approach: Find middle, reverse second half, and merge alternatingly.

from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        if not head or not head.next:
            return
        
        # Find middle
        slow, fast = head, head
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next
        
        # Reverse second half
        second = slow.next
        slow.next = None
        prev = None
        while second:
            nxt = second.next
            second.next = prev
            prev = second
            second = nxt
        
        # Merge two halves
        p, q = head, prev
        while q:
            pn, qn = p.next, q.next
            p.next = q
            q.next = pn
            p, q = pn, qn


def print_list(head):
    vals = []
    while head:
        vals.append(str(head.val))
        head = head.next
    print(' '.join(vals))


if __name__ == "__main__":
    # Test 1
    h = ListNode(1, ListNode(2, ListNode(3, ListNode(4))))
    Solution().reorderList(h)
    print_list(h)

    # Test 2
    h = ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5)))))
    Solution().reorderList(h)
    print_list(h)

"""
Time Complexity: O(n)
Space Complexity: O(1)
"""
